masschallenge/impact-api
# AC-5981: Resolve repeated queries stemming from JudgeCriteria lookups

#### Changes introduced in [AC-5981](https://masschallenge.atlassian.net/browse/AC-5981):
- remove repeated queries
- use new logic designed to get aggregates for analysis

#### How to test:

- check out AC-5981 on `impact` and `django-accelerator`
- ensure all your servers are running
- Download [this dump](https://s3.console.aws.amazon.com/s3/object/db-dump-sante-test/analysis-dump-2019-01-15.sql.gz?region=us-east-1&tab=overview&prefixSearch=ana) from S3 that has a judging round's commitments all made INCOMPLETE
- load it locally with `make load-db gz_file=<path-to-gz-dump>`

##### To validate the analysis data
- run `make django-shell` in accelerate and do the following
  - `from mc.models import *`
  - `d = User.objects.get(email__startswith="ahokie@")` (alternatively, use your email address here and give yourself the clearance needed)
  - `Clearance.objects.create(user=d, level=CLEARANCE_LEVEL_GLOBAL_MANAGER, program_family=ProgramFamily.objects.first())` (if this fails, do the following)
  - `Clearance.objects.filter(user=d, program_family_id=1).update(level=CLEARANCE_LEVEL_GLOBAL_MANAGER)`
  - `d.is_staff = d.is_superuser = True`
  - `d.save()`
  - `scenarios = Scenario.objects.filter(judging_round=48, is_active=True)`
  - `jpa = JudgeApplicationFeedback.objects.filter(panel__judgepanelassignment__scenario__in=scenarios).all()`
- at this point you can start doing interesting things e.g.
  - `first_ten = jpa[0:10]`
  - `for item in first_ten:`
  - `item.feedback_status = 'COMPLETE'`
  - `item.save()`

- if you gave `ahokie` clearances, you will have to mask as him i.e. http://localhost:8181/admin/simpleuser/user/60464/masquerade/, otherwise log in as the user you gave clearances to
- then go to the allocator at `localhost:1234/allocator` and select `2017-04 BOS CH IL Round 1` from the judging round dropdown in the top left
- notice that we have 10 reads among other analyses displayed in the gender, role, program and industry sections.
- to undo the reads, you could run the above loop and set `feedback_status` to `INCOMPLETE`

##### To see duplicate query comparisons
- from `web/impact/impact/v1/views/analyze_judging_round_view.py` do the following
  - put the following import at the top `from django.db import connection`
  - then put this in the get method before returning the response --> `print(connection.queries)`
  - this will print all the queries executed in the console, which can then be moved to a file for analysis
  - I did all this and expect we'll all have similar results. The results are the following which you can download. [Queries](https://s3.amazonaws.com/db-dump-sante-test/AC-5981/old_queries.json) on development, [queries](https://s3.amazonaws.com/db-dump-sante-test/AC-5981/new_queries.json) on AC-5981.
- to see how many times each query is executed, run
  - `sort <path-to-file> | uniq -c > [old/new]_query_count.json`
- to see the queries that have been repeated, run
  - `sort <path-to-file> | uniq -c | grep -v '^ *1 ' > [old/new]_duplicate_query_count.json`
- comparing these files will show you that we have reduced the number of duplicated queries

##### Alternate approach to review:
- check out development and hit http://localhost:8000/api/v1/analyze_judging_round/48/ 
- check out this branch, open a new tab, and hit that url again
- compare the json
- compare the queries in the django debug toolbar (toolbar does not currently show the actual query, but it does show what tables the query is hitting, what it's retrieving, and where it's called, which is generally sufficient to understand what's happening)

#### Note:
- As a bonus, we also have gender statistics now.
- the numbers that come from the old and new logic defer in the industry analyses but the results on this branch have been confirmed as correct.
